coll/han: Add alltoall algorithm

lrbison opened the pull request:
Add two Alltoall algorithms to coll/han.  Both algorithms use the same
  communication pattern.  Each rank on one host is assigned a single
  partner on a remote host and vice versa.  Then the rank collects all
  the data its partner will need to receive from it's host, and sends it
  in one large send, and likewise receives it's data in one large recv,
  then cycles to the next host.

  The two algorithms are:
  - mca_coll_han_alltoall_using_allgather: gathering data is done once
    and each rank has a copy of all local data.  Only recommended for
    small message sizes.
  - mca_coll_han_alltoall_using_smsc: ranks use smsc module to
    direct-map local memory before copying into a packed send buffer.
    Currently only the XPMEM-based smsc module supports this operation.

lrbison commented:
Some motivating data collected on 2K ranks (32 hpc7g and hpc6id instances):

![image](https://github.com/open-mpi/ompi/assets/108305344/1457c940-5ad9-4385-b4c3-5fd72168356a)


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/lrbison/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/lrbison/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{mso-number-format:"0\.0";}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


  | Han-smsc | Tuned
-- | -- | --
1 | 302.2 | 372.5
2 | 308.9 | 441.5
4 | 308.4 | 436.5
8 | 331.9 | 666.0
16 | 351.9 | 814.5
32 | 333.3 | 1703.2
64 | 435.5 | 15613.4
128 | 1248.6 | 27603.3
256 | 1866.1 | 31895.6
512 | 3421.5 | 37748.5
1024 | 6306.1 | 30685.0
2048 | 12603.2 | 38110.2
4096 | 24445.4 | 37636.9
8192 | 50277.3 | 47228.5
16384 | 97058.5 | 93082.9
32768 | 192255.5 | 185665.3
65536 | 392959.9 | 370929.0
131072 | 795386.1 | 741670.5
262144 | 1592377.8 | 1531173.9



</body>

</html>

devreal commented on ompi/mca/coll/han/coll_han.h:
```diff
@@ -486,4 +508,18 @@ ompi_coll_han_reorder_gather(const void *sbuf,
+static inline struct mca_smsc_endpoint_t *mca_coll_han_get_smsc_endpoint (struct ompi_proc_t *proc) {
+    if (OPAL_UNLIKELY(NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC])) {
+        /* LAR TODO: should we make a lock for this? */
+        /* OPAL_THREAD_LOCK(&XXX); */
+        if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
+            proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC] = mca_smsc->get_endpoint(&proc->super);
+        }
+        /* OPAL_THREAD_UNLOCK(&XXX); */
```
I think there should be a lock with a double check. Something like this:

```suggestion
        if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
            OPAL_THREAD_LOCK(&XXX);
            if (NULL == proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC]) {
                proc->proc_endpoints[OMPI_PROC_ENDPOINT_TAG_SMSC] = mca_smsc->get_endpoint(&proc->super);
            }
            OPAL_THREAD_UNLOCK(&XXX);
        }
```

lrbison replied:
I stole this code from [bml](https://github.com/open-mpi/ompi/blob/42e1cd871ccde6f625c01562818cf2ba02ac7faf/ompi/mca/bml/base/base.h#L66) which just uses a bml lock.  I suppose I could do the same for an smsc lock, I'll just need to create one.

lrbison replied:
@devreal I implemented this with a new lock scoped to the han module.

Although currently han is the only module to use the smsc endpoint, in the future this might not be true.  I feel like the lock should really be in ompi_proc_t and could apply to all get_endpoint calls, but this pattern isn't followed elsewhere so I'm hesitant to introduce it, so here we are with a new han lock.

Thoughts?

devreal commented on ompi/mca/coll/han/coll_han_alltoall.c:
```diff
@@ -0,0 +1,496 @@
+    int aint_size;
+    int bytes_to_gather;
+    int reg_data_size = 0;
+    PMPI_Type_size( MPI_AINT, &aint_size);
```
Why not just `sizeof(MPI_Aint)` here?

devreal reviewed (commented):
LGTM, just two minor comments

juntangc commented on ompi/mca/coll/han/coll_han_alltoall.c:
```diff
@@ -0,0 +1,496 @@
+        goto error;
+    }
+
+    for (int jloop=0; jloop<nrounds+fanout; jloop++) {
```
How much performance benefit by overlapping the intranode and internode communication?

lrbison replied:
Not as much as I'd hoped.  My goal wasn't so much to overlap internode and intranode, but rather pipeline internode communication.  You commented on the "using allgather" variant which doesn't perform well in large message sizes.  My observation was that for large message sizes in the smsc variant having two or perhaps three rounds in-flight at once saved me a little bit of time.  However there is an additional penalty I had to pay.  you'll notice for fanout=1 in the smsc variant I don't allocate a send_bounce buffer at all, and just re-use the application buffer.  This sidesteps the need for transport to perform ibv_mr_reg on a temporary buffer, which might change every call.  By re-using the application buffer the mr_reg cache happily re-uses previous registrations everytime and this typically outweighs the fanout improvements.

If we had an ability to re-use a buffer pool of registered memory, the fanout might be more helpful.

juntangc commented on ompi/mca/coll/han/coll_han_alltoall.c:
```diff
@@ -0,0 +1,496 @@
+    }
+
+    /* wait for all irecv to complete */
+    ompi_request_wait_all(inter_recv_count, inter_recv_reqs, MPI_STATUS_IGNORE);
```
need return and free buffer here.

wenduwan commented:
@lrbison On a high level - I am a bit bothered by the fact that we are implementing point-to-point communication directly in HAN. It breaches the encapsulation in HAN where other collective algorithms are reusing other coll modules. In other words, HAN implicitly composes hierarchical collectives of basic collectives.

Related to encapsulation, I have been thinking about the the benefits of this new algorithm. I think the biggest innovation is to separate inter-node from intra-node comms which significantly reduces network load. The intra-node offload is only a secondary concern, and we might get away with PML instead of smsc. IMO the new XHC module is a better home for the smsc stuff, and later we can adopt XHC in HAN.

lrbison commented on ompi/mca/coll/han/coll_han_alltoall.c:
```diff
@@ -0,0 +1,496 @@
+    }
+    /* wait for all irecv to complete */
+    ompi_request_wait_all(up_size, inter_recv_reqs, MPI_STATUS_IGNORE);
+
```
I've renamed the error label to cleanup to make it more clear.  After the wait_all we fall through below and free there.

lrbison commented:
@wenduwan Yes, I understand your concern about the encapsulation.  The problem is that alltoall is not easily/efficiently composed of smaller operations.  By the nature of ALL to all, we can just hierarchically subdivide the problem.

I did think of three ways to subivide:
 - each local node does allgather, then sends some portion of the data off-node.  This is implemented in this PR as alltoall_using_allgather, but it's a terrible idea for anything but the smallest message sizes due to the amount of wasted data movement (and extra memory buffers).
 - similar to above, but use gather.  We don't waste data now, but now we have to do many many gathers for each remote node.  Depending on the configuration, you probably also need a scatter on the remote side, and some transpose operation somewhere. This is also slow.

In both of those cases above the thing limiting performance is actually the local communication. Finally, there is one more case:

 - alltoall can be composed of smaller, local, alltoallv operations.  This has the potential to be the right way forward however there are two problems: (1) an efficient, smsc-based alltoallv, hasn't been implemented yet (I plan on doing so using the pattern I've written above), and (2) the setup for any smsc-based collective requires sharing the base address of all rank's buffers.  In my implementation here we pay that cost up front by an on-node allgather call and re-use the buffer address for each "round" of network communication.  However, if we call alltoallv independently at each round, then the alltoallv will have to pay that cost every time.  Conversely if we shove all communication into one round, we require very large temporary memory buffers, and it starts to look more like the allgather version.

lrbison commented:
Looks like CI has a real failure in test_cco_buf_inter.TestCCOBufInter.testAlltoall.  This is particularly surprising to me since I doubt CI setup has XPMEM installed, so I would expect it to just fall back to previous alltoall algorithm.

I will investigate, and also re-try the test including xpmem on my own cluster.

lrbison commented:
The failure is a symptom of the OB1 PML code's recent problems in INTER communicators.  Unrelated to this change.  See Issue 12407.  Ignoring that unrelated problem for a moment, I think this change is ready for review again.

wenduwan commented:
About smsc - I remember that it does not always work for accelerator, so should we add a check to unselect this algorithm for device memory? It might be tricky since it requires another consensus within the global communicator since some rank might be on accelerator but others not.

I was debating with myself if we should leave smsc out of this PR - would that cause a dramatic performance penalty? In other words, is smsc essential to achieve a reasonable/significant improvement over the current alltoall?

wenduwan commented:
> The primary advantage of this PR is that we avoid an extra copy
I have a different take on this. In my opinion the greatest value proposition of this PR is that we avoid the expensive cost associated with high network incast and outcast. In the context of HAN, every collective will benefit from smsc but I doubt if that is the foremost advantage. I imagine we can claim a bigger win if we can devise a reusable smsc pattern for allreduce, bcast, etc. too.

lrbison commented:
I realized I just sidestepped all intermediary datatypes and their conversion in this PR which is going to cause lots of failures.  I'm marking this as draft while I fix that.  I will:
 - Have followers pack data into leader buffer (which is reverse of current situation).
 - Add synchronization to account for the fact that leader doesn't know when followers are done packing.

lrbison commented:
I have completed a significant update to this PR.  The changes corrected glaring issues for DDT and device memory, while largely retaining the original performance.

 - The function properly uses convertors now, and properly detects non-contiguous and device-based memory.
 - In cases where packing is required, each low comm will decide collectively to pack their local data types into an SMSC-exposed bounce buffer of other low rank.  This however requires a few extra barriers, so it is slower.
 - In cases where packing is not required, the function uses the algorithm which I posted in this PR originally, where each rank packs data from SMSC-exposed send buffers of other low ranks into it's own bounce buffer.  This is the fast-path.
 - I removed alltoall_using_allgather.  The performance was never better than the smsc version, and only rarely better than tuned, and only for small messages.  For this reason I decided it was easier to remove than fix up the convertor usage with no benefit.

I'm marking the PR ready for review again.  Please take a look.

lrbison commented:
Updated latency results on 8x hpc7g on AWS with libfabric/EFA:

![image](https://github.com/open-mpi/ompi/assets/108305344/90265c01-de2f-4316-b955-ac8ab55990c8)

lrbison commented on ompi/mca/coll/han/coll_han_component.c:
```diff
@@ -392,6 +398,31 @@ static int han_register(void)
+                                              &cs->han_alltoall_subdivfactor);
+    cs->han_alltoall_algorithm = 0;
+    (void) mca_base_component_var_register(c, "alltoall_algorithm",
+                                              "Algorithm selection. 0=auto-based-on-size, 1=allgather-based, 2=xpmem-based.",
```
I removed the allgather-based algorithm, but a forgot to update the documentation.  Note to myself here!

lrbison replied:
Fixed.  This whole configuration block shouldn't have existed anyways, since I should have used `coll_han_use_alltoall_algorithm` instead.

edgargabriel commented:
I collected some performance numbers on one of our platforms for the alltoall_using_smsc algorithm, and compared to whatever the default is when we exclude han ( I think it is tuned, but I didn't verify). The setup is 4 nodes, 48 processes per node (192 total), Slingshot11 with libfabric 1.20.1, and XPMEM. I collected numbers when using mtl/ofi (in which case even intra-node communication will go through the cxi `provider` on this platform) and with btl/ofi + sm (btl_ofi_mode = 2).

```
# OSU MPI All-to-All Personalized Exchange Latency Test v7.2
# Datatype: MPI_CHAR.
# Size       Avg Latency(us)
#         mtl/ofi(cxi)         mtl/ofi(cxi)    btl/ofi(cxi) +sm     btl/ofi(cxi) + sm
#         han w/               w/o han         han w/               w/o han
#        alltoall_using_smsc        -          alltoall_using_smsc   -
1                     100.92        271.86         82.11             263.98
2                      98.27	    274.35         82.50	     268.71
4                      98.58	    305.49         90.40	     273.53
8                     109.88	    299.68         90.48	     275.52
16                    112.29	    299.36         91.08	     277.94
32                    113.56	    302.20         92.22	     282.38
64                    114.49	    334.46         94.11	     349.93
128                   132.85	    417.54        141.01	     405.80
256                   161.04	    646.71        169.68	     484.11
512                   243.60	   1105.65        236.03	     708.03
1024                  394.66	    465.13        357.29	     568.11
2048                  682.63	    812.04        578.92	    5030.75
4096                 1288.69	   1436.90       1030.49	    3613.06
8192                 2631.71	   3248.87       1893.00	    2856.04
16384                5336.48	   5462.84       3739.46	    5112.37
32768               11335.81	  10098.63       8581.11	    9556.01
65536               23813.99	  18430.23      22019.84	   19049.81
131072              51259.17	  36470.53      42873.01	   38622.50
262144             112964.87	  79266.76      91327.69	   67715.78
```
 I think the numbers look good overall for the new algorithm, but I can't yet make sense on why the btl/ofi performs better than with mtl/ofi for this particular scenario (where the bulk of the intra-node communication would probably go through the smsc/xpmem component in the han algorithm).

lrbison commented:
I have addressed all outstanding comments.  Thank you for the testing by @edgargabriel  and @hppritcha!  I made only small configuration-parameter related changes in my latest push:
 - I removed an unused parameter that I introduced early in my development.
 - I removed an unused `alltoall_algorithm` parameter, which was unused anyways, since `coll_han_use_alltoall_algorithm` is already doing this for me.
 - I added a trivial fallback "algorithm" so that the user can select it easily in case performance regression is observed

Please take a look.  I'm eager to merge this as it has been open a while, and I have a han alltoallv PR coming soon which builds on it.

lrbison commented:
This most recent push fixed two issues that were revealed in further testing:
 - An embarrassing refactor error meant that at some point my sends were not the correct size, as I was omitting the count (!)
 - A logic error that was related to the "push" mode which is activated when sendbuf is non-contiguous or on device-memory.

lrbison commented:
Any last comments or reviews?  This PR has been sitting for a while.

wenduwan commented:
Running AWS CI. Hopefully for the last time.
